ledger-service/http-json/src/main/scala/com/digitalasset/http/Endpoints.scala

```diff
@@ -77,17 +79,84 @@ class Endpoints(
   private def toRoute(res: => Future[Error \/ SearchResult[Error \/ JsValue]]): Route =
     responseToRoute(httpResponse(res))
 
+  // Always put this directive after a path to ensure
+  // that you don't log request bodies multiple times (simply because a matching test was made multiple times).
+  // TL;DR JUST PUT THIS THING AFTER YOUR FINAL PATH MATCHING
+  def logJsonRequestAndResult(implicit
+      lc: LoggingContextOf[InstanceUUID with RequestID]
+  ): Directive0 =
+    extractRequest & extractClientIP tflatMap { case (request, remoteAddress) =>
+      def logWithHttpMessageBodyIfAvailable(
+          httpMessage: HttpMessage,
+          msg: String,
+          bodyKind: String,
+      ): Future[Unit] =
+        if (
+          httpMessage
+            .header[`Content-Type`]
+            .map(_.contentType)
+            .contains(ContentTypes.`application/json`)
+        )
+          httpMessage
+            .entity()
+            .toStrict(maxTimeToCollectRequest)
+            .map(_.data.utf8String)
+            .map(_.parseJson)
```

Comment:
yes this can be quite the performance bottleneck as it's hella inefficient for the response body (which has been constructed by us which we are now parsing again :facepalm:), however I consider this to be a debugging helper.